src/renderer/api/kube-watch-api.ts

```diff
@@ -1,201 +1,335 @@
+    return this.apis.length > 0;
+  }
+
+  @computed get apis(): string[] {
+    const { cluster, namespaceStore } = this;
+    const activeApis = Array.from(this.subscribers.keys());
+
+    return activeApis.map(api => {
+      if (!cluster.isAllowedResource(api.kind)) {
+        return [];
+      }
+
+      if (api.isNamespaced) {
+        return namespaceStore.getContextNamespaces().map(namespace => api.getWatchUrl(namespace));
+      } else {
+        return api.getWatchUrl();
+      }
+    }).flat();
+  }
 
   constructor() {
-    reaction(() => this.activeApis, () => this.connect(), {
+    this.init();
+  }
+
+  private async init() {
+    const { getHostedCluster } = await import("../../common/cluster-store");
+    const { namespaceStore } = await import("../components/+namespaces/namespace.store");
+
+    await namespaceStore.whenReady;
+
+    this.cluster = getHostedCluster();
+    this.namespaceStore = namespaceStore;
+    this.bindAutoConnect();
+  }
+
+  private bindAutoConnect() {
+    const connect = debounce(() => this.connect(), 1000);
+
+    reaction(() => this.apis, connect, {
       fireImmediately: true,
-      delay: 500,
+      equals: comparer.structural,
     });
-  }
 
-  @computed get activeApis() {
-    return Array.from(this.subscribers.keys());
+    window.addEventListener("online", () => this.connect());
+    window.addEventListener("offline", () => this.disconnect());
+    setInterval(() => this.connectionCheck(), 60000 * 5); // every 5m
   }
 
   getSubscribersCount(api: KubeApi) {
     return this.subscribers.get(api) || 0;
   }
 
-  subscribe(...apis: KubeApi[]) {
+  subscribeApi(api: KubeApi | KubeApi[]) {
+    const apis: KubeApi[] = [api].flat();
+
     apis.forEach(api => {
       this.subscribers.set(api, this.getSubscribersCount(api) + 1);
     });
 
-    return () => apis.forEach(api => {
-      const count = this.getSubscribersCount(api) - 1;
+    return () => {
+      apis.forEach(api => {
+        const count = this.getSubscribersCount(api) - 1;
 
-      if (count <= 0) this.subscribers.delete(api);
-      else this.subscribers.set(api, count);
-    });
+        if (count <= 0) this.subscribers.delete(api);
+        else this.subscribers.set(api, count);
+      });
+    };
   }
 
-  // FIXME: use POST to send apis for subscribing (list could be huge)
-  // TODO: try to use normal fetch res.body stream to consume watch-api updates
-  // https://github.com/lensapp/lens/issues/1898
-  protected async getQuery() {
-    const { namespaceStore } = await import("../components/+namespaces/namespace.store");
+  subscribeStores(stores: KubeObjectStore[], options: IKubeWatchSubscribeStoreOptions = {}): () => void {
+    const { preload = true, waitUntilLoaded = true } = options;
+    const loading: Promise<any>[] = [];
+    const disposers: Function[] = [];
+    let isDisposed = false;
 
-    await namespaceStore.whenReady;
-    const { isAdmin } = getHostedCluster();
+    async function subscribe() {
```

Comment:
This is for a reason, because loading is happens in `async` and (un)subscribing is `sync`.
For large clusters when you visit a page you might go another one before even loading all stores and unsubscribing must happen in sync with component unmount.

Reply:
Could we instead pass in an `AbortController` which aborts the subscribes if they haven't finished yet?

Reply:
this all refactored, please re-review again